File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/ActiveInputSources.java
package io.appium.espressoserver.lib.helpers.w3c.state;

import java.util.Map;
import java.util.WeakHashMap;

import javax.annotation.Nullable;

import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException;
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource;

/**
 * Active Input Source defined in W3C spec
 *
 * (see https://www.w3.org/TR/webdriver/#terminology-0)
 */
public class ActiveInputSources {
    private final Map<String, InputSource> inputSources = new WeakHashMap<>();
    private static final Map<String, ActiveInputSources> activeInputSources = new WeakHashMap<>();

    public void addInputSource(InputSource inputSource) throws InvalidArgumentException {
        inputSources.put(inputSource.getId(), inputSource);
    }

    /**
     * Remove an input source and also remove it from InputStateTable
     * @param inputSource Source to remove
     */
    public void removeInputSource(InputSource inputSource) {
        removeInputSource(inputSource.getId());
    }

    public void removeInputSource(String id) {
        inputSources.remove(id);
    }

    @Nullable
    public InputSource getInputSource(InputSource inputSource) {
        return inputSources.get(inputSource.getId());
    }

    @Nullable
    public InputSource getInputSource(String id) {
        return inputSources.get(id);
    }

    public boolean hasInputSource(String id) {
        return inputSources.containsKey(id);
    }

    /**
     * Get the `active input sources` table for a session
     *
     * @return Global instance of ActiveInputSources
     */
    public synchronized static ActiveInputSources getActiveInputSourcesForSession(String sessionId) {
        ActiveInputSources globalInputStateTable = activeInputSources.get(sessionId);
        if (globalInputStateTable == null) {
            activeInputSources.put(sessionId, new ActiveInputSources());
            globalInputStateTable = activeInputSources.get(sessionId);
        }
        return globalInputStateTable;
    }
}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/ActiveInputSources.kt
package io.appium.espressoserver.lib.helpers.w3c.state

import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource
import java.util.*

/**
 * Active Input Source defined in W3C spec
 *
 * (see https://www.w3.org/TR/webdriver/#terminology-0)
 */
class ActiveInputSources {
    private val inputSources: MutableMap<String?, InputSource> = WeakHashMap()

    @Throws(InvalidArgumentException::class)
    fun addInputSource(inputSource: InputSource) {
        inputSources[inputSource.id] = inputSource
    }

    /**
     * Remove an input source and also remove it from InputStateTable
     * @param inputSource Source to remove
     */
    fun removeInputSource(inputSource: InputSource) {
        removeInputSource(inputSource.id)
    }

    fun removeInputSource(id: String?) {
        inputSources.remove(id)
    }

    fun getInputSource(inputSource: InputSource): InputSource? {
        return inputSources[inputSource.id]
    }

    fun getInputSource(id: String?): InputSource? {
        return inputSources[id]
    }

    fun hasInputSource(id: String?): Boolean {
        return inputSources.containsKey(id)
    }

    companion object {
        private val activeInputSources: MutableMap<String, ActiveInputSources> = WeakHashMap()

        /**
         * Get the `active input sources` table for a session
         *
         * @return Global instance of ActiveInputSources
         */
        @Synchronized
        fun getActiveInputSourcesForSession(sessionId: String): ActiveInputSources {
            var globalInputStateTable = activeInputSources[sessionId]
            if (globalInputStateTable == null) {
                activeInputSources[sessionId] = ActiveInputSources()
                globalInputStateTable = activeInputSources[sessionId]
            }
            return globalInputStateTable!!
        }
    }
}